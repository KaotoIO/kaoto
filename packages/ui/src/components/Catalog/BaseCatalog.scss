.catalog-list {
  overflow: auto;
  padding: var(--pf-v5-global--spacer--sm);
}
