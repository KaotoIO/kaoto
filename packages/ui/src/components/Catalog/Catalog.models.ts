export interface ITile {
  type: string;
  name: string;
  title: string;
  description?: string;
  headerTags?: string[];
  tags: string[];
  version?: string;
  provider?: string;
}

export type TileFilter = (item: ITile) => boolean;

export const enum CatalogLayout {
  Gallery = 'Gallery',
  List = 'List',
}
