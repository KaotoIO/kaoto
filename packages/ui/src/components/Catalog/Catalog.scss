.catalog {
  overflow: inherit;
  max-height: 100%;
  display: flex;
  flex-direction: column;

  &__filter {
    margin-bottom: var(--pf-t--global--spacer--md);
  }

  &__base {
    margin-bottom: var(--pf-t--global--spacer--md);
  }
}
