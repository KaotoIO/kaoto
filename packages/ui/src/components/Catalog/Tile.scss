.tile {
  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v5-c-card__actions {
    padding: 0;
    margin: 0;
  }

  &__header,
  &__footer {
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-flow: wrap;
  }

  &__icon {
    width: 25px;
  }

  &__name {
    text-align: center;
    min-height: 25px;
    font-size: small;
    display: flex;
    flex-flow: column;
    align-items: center;
  }

  &__title {
    text-align: center;
    display: flex;
    flex-flow: column;
    align-items: center;
  }

  &__body {
    max-height: 150px;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &__provider {
    font-size: var(--pf-v5-c-card__body--FontSize);
    text-align: right;
    padding: 0 10px;
    font-weight: bold;
    background-color: var(--pf-v5-global--palette--black-150);
  }
}
