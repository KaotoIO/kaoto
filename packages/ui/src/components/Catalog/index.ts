export * from './Catalog';
export * from './CatalogLayoutIcon';
export * from './Catalog.models';
export * from './DataListItem';
export * from './Tile';
export * from './tag-color-resolver';
