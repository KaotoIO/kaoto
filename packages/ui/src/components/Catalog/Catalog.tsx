import { FunctionComponent, PropsWithChildren, useCallback, useMemo, useState } from 'react';
import { useDebounceValue } from 'usehooks-ts';
import { useLocalStorage } from '../../hooks';
import { LocalStorageKeys } from '../../models';
import { BaseCatalog } from './BaseCatalog';
import { CatalogLayout, ITile } from './Catalog.models';
import './Catalog.scss';
import { CatalogFilter } from './CatalogFilter';
import { filterTiles } from './filter-tiles';
import { sortTags } from './sort-tags';

interface CatalogProps {
  /** Tiles list */
  tiles: ITile[];

  /** Callback being called when a Tile is clicked */
  onTileClick?: (tile: ITile) => void;
}

export const Catalog: FunctionComponent<PropsWithChildren<CatalogProps>> = (props) => {
  const [activeLayout, setActiveLayout] = useLocalStorage(LocalStorageKeys.CatalogLayout, CatalogLayout.Gallery);

  /** Selected Group */
  const [searchTerm, setSearchTerm] = useDebounceValue('', 500, { trailing: true });
  const [filterTags, setFilterTags] = useState<string[]>([]);

  /** All tags, sorted to have selected and prioritized tags first */
  const { sortedTags, overflowIndex } = useMemo(() => {
    return sortTags(props.tiles, filterTags);
  }, [filterTags, props.tiles]);

  /** All providers */
  const providers = useMemo(() => {
    const providersSet = new Set<string>();
    props.tiles.forEach((tile) => {
      providersSet.add(tile.provider ?? 'Community');
    });

    return Array.from(providersSet);
  }, [props.tiles]);

  /** Selected Providers */
  const [selectedProviders, setSelectedProviders] = useState<string[]>(providers);

  const filteredTiles = useMemo(() => {
    return filterTiles(props.tiles, { searchTerm, searchTags: filterTags, selectedProviders });
  }, [filterTags, props.tiles, searchTerm, selectedProviders]);

  /** Set the tiles groups */
  const tilesGroups = useMemo(() => {
    const groups: Record<string, ITile[]> = {};
    filteredTiles.forEach((tile) => {
      if (!groups[tile.type]) {
        groups[tile.type] = [];
      }
      groups[tile.type].push(tile);
    });
    return Object.entries(groups).map(([group, tiles]) => ({ name: group, count: tiles.length }));
  }, [filteredTiles]);

  const [activeGroups, setActiveGroups] = useState<string[]>(tilesGroups.map((g) => g.name));

  const filteredTilesByGroup = useMemo<ITile[]>(() => {
    return filteredTiles.filter((tile) => activeGroups.includes(tile.type));
  }, [activeGroups, filteredTiles]);

  const onFilterChange = useCallback(
    (_event: unknown, value = '') => {
      setSearchTerm(value);
    },
    [setSearchTerm],
  );

  const onTileClick = useCallback(
    (tile: ITile) => {
      props.onTileClick?.(tile);
    },
    [props],
  );

  const onTagClick = useCallback((_event: unknown, value = '') => {
    setFilterTags((previousFilteredTags) => {
      return previousFilteredTags.includes(value) ? previousFilteredTags : [...previousFilteredTags, value];
    });
  }, []);

  const onSelectProvider = useCallback(
    (provider: string) => {
      setSelectedProviders(
        selectedProviders.includes(provider)
          ? selectedProviders.filter((selection) => selection !== provider)
          : [provider, ...selectedProviders],
      );
    },
    [selectedProviders],
  );

  return (
    <div className="catalog">
      <CatalogFilter
        className="catalog__filter"
        searchTerm={searchTerm}
        groups={tilesGroups}
        providers={providers}
        selectedProviders={selectedProviders}
        tags={sortedTags}
        tagsOverflowIndex={overflowIndex}
        layouts={[CatalogLayout.Gallery, CatalogLayout.List]}
        activeGroups={activeGroups}
        activeLayout={activeLayout}
        filterTags={filterTags}
        onChange={onFilterChange}
        setActiveGroups={setActiveGroups}
        setActiveLayout={setActiveLayout}
        setFilterTags={setFilterTags}
        onSelectProvider={onSelectProvider}
      />
      <BaseCatalog
        className="catalog__base"
        tiles={filteredTilesByGroup}
        catalogLayout={activeLayout}
        onTileClick={onTileClick}
        onTagClick={onTagClick}
      />
    </div>
  );
};
