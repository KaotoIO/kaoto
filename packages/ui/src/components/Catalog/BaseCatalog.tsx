import { DataList, Gallery, Pagination } from '@patternfly/react-core';
import { FunctionComponent, useCallback, useEffect, useRef, useState } from 'react';
import './BaseCatalog.scss';
import { CatalogLayout, ITile } from './Catalog.models';
import { CatalogDataListItem } from './DataListItem';
import { Tile } from './Tile';

interface BaseCatalogProps {
  className?: string;
  tiles: ITile[];
  catalogLayout: CatalogLayout;
  onTileClick?: (tile: ITile) => void;
  onTagClick: (_event: unknown, value: string) => void;
}

export const BaseCatalog: FunctionComponent<BaseCatalogProps> = (props) => {
  const catalogBodyRef = useRef<HTMLDivElement>(null);
  const itemCount = props.tiles?.length;
  const [page, setPage] = useState(1);
  const [perPage, setPerPage] = useState(50);

  const onTileClick = useCallback(
    (tile: ITile) => {
      props.onTileClick?.(tile);
    },
    [props],
  );

  const startIndex = (page - 1) * perPage < 0 ? 0 : (page - 1) * perPage;
  const endIndex = page * perPage;
  useEffect(() => {
    // Handling the scenario where the item count is less the page selected.
    if (startIndex + 1 > itemCount) {
      setPage(Math.ceil(itemCount / perPage));
    } else if (page === 0 && itemCount > 0) setPage(1);

    catalogBodyRef.current!.scrollTop = 0;
  }, [props.tiles]);

  const onSelectDataListItem = useCallback(
    (_event: React.MouseEvent | React.KeyboardEvent, id: string) => {
      const tile = props.tiles.find((tile) => tile.name + '-' + tile.type === id);
      onTileClick(tile!);
    },
    [onTileClick, props.tiles],
  );

  const onSetPage = (_event: React.MouseEvent | React.KeyboardEvent | MouseEvent, newPage: number) => {
    setPage(newPage);
  };

  const onPerPageSelect = (
    _event: React.MouseEvent | React.KeyboardEvent | MouseEvent,
    newPerPage: number,
    newPage: number,
  ) => {
    setPerPage(newPerPage);
    setPage(newPage);
  };

  const paginatedCards = props.tiles?.slice(startIndex, endIndex);

  return (
    <>
      <Pagination
        itemCount={itemCount}
        perPage={perPage}
        page={page}
        onSetPage={onSetPage}
        widgetId="catalog-pagination"
        onPerPageSelect={onPerPageSelect}
        ouiaId="CatalogPagination"
      />
      <div id="catalog-list" className="catalog-list" ref={catalogBodyRef}>
        {props.catalogLayout == CatalogLayout.List && (
          <DataList aria-label="Catalog list" onSelectDataListItem={onSelectDataListItem} isCompact>
            {paginatedCards.map((tile) => (
              <CatalogDataListItem key={`${tile.name}-${tile.type}`} tile={tile} onTagClick={props.onTagClick} />
            ))}
          </DataList>
        )}
        {props.catalogLayout == CatalogLayout.Gallery && (
          <Gallery hasGutter>
            {paginatedCards.map((tile) => (
              <Tile key={`${tile.name}-${tile.type}`} tile={tile} onClick={onTileClick} onTagClick={props.onTagClick} />
            ))}
          </Gallery>
        )}
      </div>
    </>
  );
};
