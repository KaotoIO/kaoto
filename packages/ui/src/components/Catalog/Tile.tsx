import { Card, CardBody, CardFooter, CardHeader, CardTitle, LabelGroup } from '@patternfly/react-core';
import { FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { IconResolver } from '../IconResolver';
import { ITile } from './Catalog.models';
import { CatalogTag, CatalogTagsPanel } from './Tags';
import './Tile.scss';

interface TileProps {
  tile: ITile;
  onClick: (tile: ITile) => void;
  onTagClick: (_event: unknown, value: string) => void;
}

export const Tile: FunctionComponent<PropsWithChildren<TileProps>> = (props) => {
  const onTileClick = useCallback(() => {
    props.onClick(props.tile);
  }, [props]);

  return (
    <Card
      className="tile"
      data-testid={'tile-' + props.tile.name}
      isClickable
      isCompact
      role="button"
      key={props.tile.name}
      id={props.tile.name}
      onClick={onTileClick}
    >
      <CardHeader
        selectableActions={{
          variant: 'single',
          selectableActionId: props.tile.name,
          selectableActionAriaLabelledby: `Selectable ${props.tile.name}`,
          name: props.tile.name,
        }}
      >
        <div className="tile__header">
          <IconResolver className="tile__icon" tile={props.tile} />
          <LabelGroup isCompact aria-label="tile-headers-tags">
            {props.tile.headerTags?.map((tag, index) => (
              <CatalogTag key={`${props.tile.name}-${tag}-${index}`} tag={tag} />
            ))}
          </LabelGroup>
        </div>

        <CardTitle className="tile__title">
          <span>{props.tile.title}</span>
          {props.tile.version && (
            <CatalogTag key={`${props.tile.version}`} tag={props.tile.version} variant="outline" />
          )}
        </CardTitle>
      </CardHeader>

      <CardBody className="tile__body">{props.tile.description}</CardBody>

      <CardFooter>
        <CatalogTagsPanel tags={props.tile.tags} onTagClick={props.onTagClick} />
      </CardFooter>
    </Card>
  );
};
