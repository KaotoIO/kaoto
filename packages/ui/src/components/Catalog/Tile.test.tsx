import { fireEvent, render } from '@testing-library/react';
import { Tile } from './Tile';
import { ITile } from './Catalog.models';

describe('Tile', () => {
  const tile: ITile = {
    type: 'tile-type',
    name: 'tile-name',
    title: 'tile-title',
    description: 'tile-description',
    tags: ['tag1', 'tag2'],
    headerTags: ['header-tag1', 'header-tag2'],
  };

  it('renders correctly', () => {
    const { container } = render(<Tile tile={tile} onClick={jest.fn()} onTagClick={jest.fn()} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls onClick and onTagClick prop when clicked', () => {
    const onClick = jest.fn();
    const onTagClick = jest.fn();

    const { getByTestId } = render(<Tile tile={tile} onClick={onClick} onTagClick={onTagClick} />);

    fireEvent.click(getByTestId('tile-header-tile-name'));
    fireEvent.click(getByTestId('tag-tag1'));

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onTagClick).toHaveBeenCalledTimes(1);
  });
});
