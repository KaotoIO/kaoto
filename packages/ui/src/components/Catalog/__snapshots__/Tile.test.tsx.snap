// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders correctly 1`] = `
<div
  class="pf-v5-c-card pf-m-compact pf-m-clickable tile"
  data-ouia-component-id="OUIA-Generated-Card-1"
  data-ouia-component-type="PF5/Card"
  data-ouia-safe="true"
  data-testid="tile-tile-name"
  id="tile-name"
  role="button"
>
  <div
    class="pf-v5-c-card__header"
    data-testid="tile-header-tile-name"
  >
    <div
      class="pf-v5-c-card__actions"
    >
      <div
        class="pf-v5-c-card__selectable-actions"
      >
        <div
          class="pf-v5-c-radio pf-m-standalone"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selectable tile-name"
            class="pf-v5-c-radio__input pf-v5-screen-reader"
            data-ouia-component-id="OUIA-Generated-Radio-1"
            data-ouia-component-type="PF5/Radio"
            data-ouia-safe="true"
            id="tile-name"
            name="tile-name"
            type="radio"
          />
          <label
            class="pf-v5-c-radio__label"
            for="tile-name"
          />
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-card__header-main"
    >
      <div
        class="tile__header"
      >
        <img
          alt="camel icon"
          class="tile__icon"
          src=""
        />
        <div
          class="pf-v5-c-label-group"
        >
          <div
            class="pf-v5-c-label-group__main"
          >
            <ul
              aria-label="tile-headers-tags"
              class="pf-v5-c-label-group__list"
              role="list"
            >
              <li
                class="pf-v5-c-label-group__list-item"
              >
                <span
                  class="pf-v5-c-label pf-m-compact"
                >
                  <span
                    class="pf-v5-c-label__content"
                  >
                    <span
                      class="pf-v5-c-label__text"
                    >
                      header-tag1
                    </span>
                  </span>
                </span>
              </li>
              <li
                class="pf-v5-c-label-group__list-item"
              >
                <span
                  class="pf-v5-c-label pf-m-compact"
                >
                  <span
                    class="pf-v5-c-label__content"
                  >
                    <span
                      class="pf-v5-c-label__text"
                    >
                      header-tag2
                    </span>
                  </span>
                </span>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-c-card__title"
      >
        <div
          class="pf-v5-c-card__title-text tile__title"
          id="tile-name-title"
        >
          <span>
            tile-title
          </span>
          <span
            class="tile__name"
          >
            (
            tile-name
            )
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="pf-v5-c-card__body tile__body"
  >
    tile-description
  </div>
  <div
    class="pf-v5-c-card__footer"
  >
    <div
      class="pf-v5-c-label-group"
    >
      <div
        class="pf-v5-c-label-group__main"
      >
        <ul
          aria-label="data-list-item-tags"
          class="pf-v5-c-label-group__list"
          role="list"
        >
          <li
            class="pf-v5-c-label-group__list-item"
          >
            <span
              class="pf-v5-c-label pf-m-compact"
            >
              <a
                class="pf-v5-c-label__content"
                data-testid="tag-tag1"
              >
                <span
                  class="pf-v5-c-label__text"
                >
                  tag1
                </span>
              </a>
            </span>
          </li>
          <li
            class="pf-v5-c-label-group__list-item"
          >
            <span
              class="pf-v5-c-label pf-m-compact"
            >
              <a
                class="pf-v5-c-label__content"
                data-testid="tag-tag2"
              >
                <span
                  class="pf-v5-c-label__text"
                >
                  tag2
                </span>
              </a>
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
